wfe/wfe.go

```diff
@@ -946,10 +947,19 @@ func (wfe *WebFrontEndImpl) Challenge(
 		return
 	}
 	authorizationID := slug[0]
-	challengeID, err := strconv.ParseInt(slug[1], 10, 64)
+	var challengeID interface{}
+	var err error
+	challengeID, err = strconv.ParseInt(slug[1], 10, 64)
 	if err != nil {
-		notFound()
-		return
+		if features.Enabled(features.NewAuthorizationSchema) {
+			// If the ID isn't an int and features.NewAuthorizationSchema is
```

Comment:
Yeah, I don't think there is really a solution to this problem. There will still be the edge case where we generate a ID which _could_ be interpreted as a valid ID for a challenge, it'd be incredibly unlikely, but still possible.

The more I think about it I think we should probably just go with the method you suggested previously which is to just generate set of static random strings and use them for the challenge IDs. It's not my favourite idea, but it's the only one I can think of at the moment that doesn't have weird corner cases 😕.

Reply:
What do you think of my proposed fix above? Once we fetch the Authorization, we know whether we're using a v1 or a v2 Authorization, and can interpret the challenge ID accordingly.

Reply:
Currently we don't have a way of telling if an authorization is v1 or v2 outside of the SA, there was some discussion previously of having some way of doing this but I don't think it was ever implemented. Adding this would definitely simplify this problem... I'd be in favor of doing that and then sticking with the current ID style here.

Reply:
Can't you check if all the Challenges have a numeric ID of 0?

Reply:
Sure, that's one way of testing if it's v2. We do already have a v2 field on the protobuf though which appears to never have been wired up to the actual core object. I'd rather just pass that through so we have a nice way to immediately check if something is v2 or not, which seems like it'll also be useful down the line for a handful of the other remaining changes.

Reply:
Although hrm, I realize now that the WFE currently only supports retrieving v1 authorizations, the logic for retrieving v2 authorizations isn't implemented yet (working on this in chunks may have actually been somewhat counterproductive...). Depending on where that logic is implemented (WFE vs. SA) you may end up needing to know what type of authz is being requested before you actually get the authorization at all (if implemented at the WFE) so knowing which kind it is for the challenge is somewhat redundant...

I'm going to think about this more tomorrow, at somewhat of an impasse at the moment.

Reply:
Want to just make it an explicit part of the URL? Like `/acme/authz/v2/1234/lwps`